ddtrace/internal/datadog/profiling/src/exporter.cpp

```diff
@@ -0,0 +1,560 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0. This product includes software
+// developed at Datadog (https://www.datadoghq.com/). Copyright 2021-Present
+// Datadog, Inc.
+
+// High-level skip for invalid architectures
+#ifndef __linux__
+#elif __aarch64__
+#elif __i386__
+#else
+
+#include "exporter.hpp"
+#include <iostream>
+
+#include <fcntl.h>
+#include <sys/stat.h>
+#include <sys/types.h>
+#include <unistd.h>
+
+using namespace Datadog;
+
+inline ddog_CharSlice to_slice(std::string_view str) {
+  return {.ptr = str.data(), .len = str.size()};
+}
+
+DdogProfExporter::DdogProfExporter(
+    std::string_view env, std::string_view service, std::string_view version,
+    std::string_view runtime, std::string_view runtime_version,
+    std::string_view profiler_version, std::string_view url,
+    ExporterTagset &user_tags) {
+
+  // Setup
+  ddog_Vec_Tag tags = ddog_Vec_Tag_new();
+  add_tag(tags, ExportTagKey::language, language);
+  add_tag(tags, ExportTagKey::env, env);
+  add_tag(tags, ExportTagKey::service, service);
+  add_tag(tags, ExportTagKey::version, version);
+  add_tag(tags, ExportTagKey::runtime, runtime);
+  add_tag(tags, ExportTagKey::runtime_version, runtime_version);
+  add_tag(tags, ExportTagKey::profiler_version, profiler_version);
+
+  // Add the unsafe tags, if any
+  for (const auto &kv : user_tags)
+    add_tag_unsafe(tags, kv.first, kv.second);
+
+  ddog_prof_Exporter_NewResult new_exporter = ddog_prof_Exporter_new(
+      to_slice("dd-trace-py"), to_slice(profiler_version), to_slice(family),
+      &tags, ddog_Endpoint_agent(to_slice(url)));
+  ddog_Vec_Tag_drop(tags);
+
+  if (new_exporter.tag == DDOG_PROF_EXPORTER_NEW_RESULT_OK) {
+    ptr = new_exporter.ok;
+  } else {
+    // TODO consolidate errors
```

Comment:
I think the builder provides an easy way to at least stop profiing if things go wrong in the exporter creation.

Reply:
Very true.  Fixed in 61f10cc50 by

1. getting rid of the wrapper class over ddog_profile_Export
2. checking subsequent initialization in the builder
3. adding a custom deleter to the unique_ptr, which is now in the uploader

still need to consume the check, but I think the immediate part is better now.